app/tor.js

```diff
@@ -0,0 +1,1253 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+// Tor daemon management.
+//
+// This doesn't actually manage the tor daemon: the parts that did are
+// commented out.  Rather it just watches the tor daemon's control
+// port file for activity and connects to its control socket.
+
+const EventEmitter = require('events')
+const assert = require('assert')
+const electron = require('electron')
+const fs = require('fs')
+const net = require('net')
+const path = require('path')
+const stream = require('stream')
+
+// torBravePath()
+//
+//      Return the path to the directory where we store tor-related
+//      files.
+//
+function torBravePath () {
+  const bravePath = electron.app.getPath('userData')
+  return path.join(bravePath, 'tor')
+}
+
+// torDataDirPath()
+//
+//      Return the path to the data directory that we use for our tor
+//      daemon.
+//
+function torDataDirPath () {
+  return path.join(torBravePath(), 'data')
+}
+
+// torWatchDirPath()
+//
+//      Return the path to the directory we watch for changes as tor
+//      starts up.
+//
+function torWatchDirPath () {
+  return path.join(torBravePath(), 'watch')
+}
+
+// torControlPortPath()
+//
+//      Return the path to the file containing the port number for the
+//      control channel that our tor daemon is listening on.
+//
+function torControlPortPath () {
+  return path.join(torWatchDirPath(), 'controlport')
+}
+
+// torControlCookiePath()
+//
+//      Return the path to the file containing the control connection
+//      authentication cookie.
+//
+function torControlCookiePath () {
+  return path.join(torDataDirPath(), 'control_auth_cookie')
+}
+
+// TorDaemon
+//
+//      State for a tor daemon subprocess.
+//
+class TorDaemon extends EventEmitter {
+  constructor () {
+    super()
+    this._process = null        // child process
+    this._watcher = null        // fs watcher
+    this._polling = false       // true if we are polling for start
+    this._retry_poll = null     // set if polling, true if should retry on fail
+    this._control = null        // TorControl instance
+    this._socks_addresses = null // array of tor's socks addresses
+  }
+
+  // setup(callback)
+  //
+  //    Create the necessary directories and invoke callback when
+  //    done.  We assume the parent of torBravePath exists; we create
+  //    it and everything we need underneath it.  On failure other
+  //    than EEXIST, may leave directories partially created.
+  //
+  setup (callback) {
+    const directories = [
+      torBravePath(),
+      torWatchDirPath()
+    ]
+    const loop = (i) => {
+      if (i === directories.length) {
+        return callback(null)
+      }
+      assert(i >= 0)
+      assert(i < directories.length)
+      fs.mkdir(directories[i], 0o700, (err) => {
+        if (err && err.code !== 'EEXIST') {
+          return callback(err)
+        }
+        loop(i + 1)
+      })
+    }
+    loop(0)
```

Comment:
it would be slightly clearer if you used `directories.forEach` instead of this recursive loop

Reply:
Callbacks make forEach a little tricky.  If this should be done with mkdirSync, I guess that's OK, but I figured it would be better to avoid that.  (I'm not _sure_ this is necessary, but the ordering of directory creation in muon vs browser-laptop is not obvious, so I figured it would be safer to keep this logic than delete it.)

Reply:
@riastradh-brave i see now that one of the directories is a subdir of the other. in that case you can just use fs-extra's mkdir instead of fs's mkdir, which will behave like `mkdir -p`: https://github.com/jprichardson/node-fs-extra

Reply:
How does that help to do multiple ones in a loop?  I guess I could replace the whole thing by a single mkdirp for the deepest subdirectory, but I didn't want to embed the assumption that there is a single deepest subdirectory in that function.